multiqc/modules/bamqc/bamqc.py

```diff
@@ -145,10 +178,159 @@ def find_outliers(self):
                                    self.bamqc_data[s_name]['statuses'][property] = 'fail'
                         else:
                                    self.bamqc_data[s_name]['statuses'][property] = 'pass'
+                        
+                # Check if (mean - 2 * stdev) < 0; if so, set the value in bamqc_stats_warning to be true
+                n_property = property.lower().replace(' ', '_')
+                if (mean - 2 * stdev) < 0:
+                        self.bamqc_stats_warning[n_property] = True
+                else:
+                        self.bamqc_stats_warning[n_property] = False
 
                 # reset the list
                 reads = [] 
 
+    def bamqc_general_stats(self):
+        """ Add single-number stats to the bamqc general statistics
+        table at the top of the report but below the general stats table"""
+
+        # Prepare the data
+        
+        # Calculate map %: 
+        # formula = (mapped reads / total reads) * 100
+        # if total reads is 0, treat as 1
+        for s_name in self.bamqc_data.keys():
+                mapped_reads = self.bamqc_data[s_name]['mapped reads']
+                total_reads = self.bamqc_data[s_name]['total reads']
+                if total_reads == 0:
+                        total_reads = 1
+                map_percent = (mapped_reads / total_reads) * 100
+                self.bamqc_data[s_name]['map percent'] = map_percent
+
+        # Calculate on target %:
+        # formula = (reads on target / mapped reads) * 100
+        # if mapped reads is 0, treat as 1
+        for s_name in self.bamqc_data.keys():
+                mapped_reads = self.bamqc_data[s_name]['mapped reads']
+                reads_on_target = self.bamqc_data[s_name]['reads on target']
+                if mapped_reads == 0:
+                        mapped_reads = 1
+                on_target_percent = (reads_on_target / mapped_reads) * 100
+                self.bamqc_data[s_name]['on target percent'] = on_target_percent
+
+        # Calculate the estimated total coverage
+        # formula = estimated yield / target size
+        # if target size is 0, treat as 1        
+        # formula for the estimated yield = (total aligned bases * on target percentage) / reads per start point
+        # if reads per start point is 0, treat as 1
+        for s_name in self.bamqc_data.keys():
```

Comment:
If you're only working with the values, it might be clearer to iterate over the values and modify the dictionary.

Reply:
Thanks for the review.
I am not sure how to do so. I need to go over each sample in the dictionary, and the sample name is the key. Do you mean that I'd better have another for loop to iterate values of each sample?